src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,373 @@
+
+        this.ReturnScopeMetrics();
+    }
+
+    internal void ReturnScopeMetrics()
+    {
+        foreach (var list in this.scopeMetrics)
+        {
+            list.Value.Clear();
+            MetricListPool.Add(list.Value);
+        }
+
+        this.scopeMetrics.Clear();
+    }
+
+    internal void SerializeResourceMetrics(byte[] buffer, ref int currentPosition, Resource resource)
+    {
+        this.resourceMetricStartIndex = currentPosition;
+
+        currentPosition += LengthAndTagSize;
+
+        // Serialize Resource
+        // TODO: Avoid serializing it multiple times.
+        SerializeResource(buffer, ref currentPosition, resource);
+
+        // TODO: Serialize schema_url field
+
+        // Serialize ScopeMetrics field
+        foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
+        {
+            this.previousScopeMetricsStartIndex = currentPosition;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize this meter/scope
+            this.SerializeScopeMetrics(buffer, ref currentPosition, metric.Key, metric.Value);
+
+            // Reset the current position to start writing again from where we started the last scope.
+            currentPosition = this.previousScopeMetricsStartIndex;
+        }
+    }
+
+    internal void SerializeScopeMetrics(byte[] buffer, ref int currentPosition, string scopeName, List<Metric> metrics)
+    {
+        bool scopeInformationIncluded = false;
+
+        foreach (Metric metric in metrics)
+        {
+            if (!scopeInformationIncluded)
+            {
+                // Serialize scope information
+                SerializeInstrumentationScope(buffer, ref currentPosition, scopeName, metric.MeterTags);
+
+                scopeInformationIncluded = true;
+            }
+
+            this.previousMetricStartIndex = currentPosition;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize metrics for the meter/scope
+            this.SerializeMetric(buffer, ref currentPosition, metric);
+
+            // Reset the current position to start writing again from where we started the last metric.
+            currentPosition = this.previousMetricStartIndex;
+        }
+
+        // TODO: Serialize schema_url field.
+    }
+
+    private void SerializeMetric(byte[] buffer, ref int currentPosition, Metric metric)
+    {
+        WriteInstrumentDetails(buffer, ref currentPosition, metric);
+
+        switch (metric.MetricType)
+        {
+            case MetricType.LongSum:
+            case MetricType.LongSumNonMonotonic:
+                {
+                    this.previousIntrumentStartIndex = currentPosition;
+
+                    currentPosition += LengthAndTagSize;
+
+                    // Write isMonotonic tag
+                    ProtobufSerializerHelper.WriteBoolWithTag(buffer, ref currentPosition, FieldNumberConstants.Sum_is_monotonic, metric.MetricType == MetricType.LongSum);
```

Comment:
Why are we serializing this in the reverse order?

From the proto:

```proto
message Sum {
  repeated NumberDataPoint data_points = 1;

  // aggregation_temporality describes if the aggregator reports delta changes
  // since last report time, or cumulative changes since a fixed start time.
  AggregationTemporality aggregation_temporality = 2;

  // If "true" means that the sum is monotonic.
  bool is_monotonic = 3;
}
```

Reply:
It's just easier this way. We need to serialize field # 2 and 3 only once so we serialize it and then keep writing metricPoint one by one. Sequence does not matter here.

Alternate way would be to serialize those fields every time as we do not know in advance how much space a metricPoint would take.